src/pages/workspaces/workspace/WorkspaceContainer.js

```diff
@@ -152,6 +152,33 @@ const WorkspaceAccessError = () => {
   ])
 }
 
+const useClusterPolling = ({ namespace }) => {
+  const signal = useCancellation()
+  const timeout = useRef()
+  const [clusters, setClusters] = useState(undefined)
+  const reschedule = ms => {
+    clearTimeout(timeout.current)
+    timeout.current = setTimeout(refreshClustersSilently, ms)
+  }
+  const loadClusters = async () => {
+    try {
+      const newClusters = await Ajax(signal).Jupyter.clustersList(namespace)
+      setClusters(_.filter({ creator: getUser().email }, newClusters))
+      const cluster = currentCluster(newClusters)
+      reschedule(_.includes(cluster && cluster.status, ['Creating', 'Starting', 'Stopping']) ? 10000 : 120000)
+    } catch (error) {
+      reschedule(30000)
+      throw error
+    }
+  }
+  const refreshClusters = withErrorReporting('Error loading clusters', loadClusters)
+  const refreshClustersSilently = withErrorIgnoring(loadClusters)
```

Comment:
Should these two be defined up at the top, so they're above `reschedule`?

Reply:
They can't be, because they reference `loadClusters` directly. There's a (deliberate) circle here, but it works as written because the forward variable access is inside the body of `reschedule`, and by the time that's actually called the variable exists.